### PR TITLE
Implement currencyContext and make version agnostic payInput component

### PR DESCRIPTION
**Pull request by johnnyd-eth**
## What does this PR do and why?

This PR is intended to reduce the diff of #611 , although is itself quite thicc. 

**Key code changes structure:**

- Makes a version agnostic PayInput component which gets passed:
      - Version-specific PayButton (`V1PayButton`)
      - Version-specific payProjectTx transaction function `payV1ProjectTx`
      
- Creates a CurrencyContext and uses it in PayInput for the currency toggle

_This PR houldn't cause any changes to the UI_ 

## Acceptance checklist

- [x] I have evaluated the [Approval Guidelines](../CONTRIBUTING.md#approval-guidelines) for this PR.
- [x] I have tested this PR in [all supported browsers](../CONTRIBUTING.md#browser-support).

### CODE REVIEW

File: src/components/shared/inputs/Pay/PayInput.tsx [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+          }
+        />
+        <PayInputSubText
+          payInCurrency={payInCurrency ?? currencyETH}
+          amount={payAmount}
+          reservedRate={reservedRate}
+          weight={weight}
+          tokenSymbol={tokenSymbol}
+          tokenAddress={tokenAddress}
+        />
```
Comment (aeolianeth) [inline src/components/shared/inputs/Pay/PayInput.tsx:72]: I guess this part is really the "PayInput". Should this be it's own component, or should we rename this component to be `PayInputButton` or `PayInputGroup` or something?

File: src/components/shared/PayWarningModal.tsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import { Modal } from 'antd'
```
Comment (aeolianeth) [inline src/components/shared/PayWarningModal.tsx:1]: **praise**

File: src/components/shared/inputs/Pay/PayInputSubText.tsx [hunk 3]
```diff
@@ -27,27 +25,40 @@ import { V1_CURRENCY_ETH } from 'constants/v1/currency'
...
+    currencyMetadata,
+    currencies: { currencyETH },
+  } = useContext(CurrencyContext)
+
   const tokenText = tokenSymbolText({
     tokenSymbol: tokenSymbol,
     capitalize: false,
     plural: true,
   })
+  const weiPayAmt = AmountToWei({ currency: payInCurrency, amount: amount })
```
Comment (aeolianeth) [inline src/components/shared/inputs/Pay/PayInputSubText.tsx:57]: **issue** i think hooks should only be within other hooks.

AmountToWei is essentially a hook here.

We could have a hook useWeiConverter that does this, perhaps.

File: src/components/v1/V1Dashboard/index.tsx [hunk 4]
```diff
@@ -232,21 +235,22 @@ export default function V1Dashboard() {
 
   return (
     <V1ProjectContext.Provider value={project}>
-      <div style={layouts.maxWidth}>
-        <V1Project />
-
-        <div style={{ textAlign: 'center', padding: 20 }}>
-          <ScrollToTopButton />
+      <CurrencyContext.Provider value={V1_CURRENCY_CONTEXT}>
```
Comment (aeolianeth) [inline src/components/v1/V1Dashboard/index.tsx:238]: juicy

File: src/components/v1/V1Project/Pay/V1PayButton.tsx [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+import { decodeFundingCycleMetadata } from 'utils/v1/fundingCycle'
+import AmountToWei from 'utils/AmountToWei'
+import { usePayV1ProjectTx } from 'hooks/v1/transactor/PayV1ProjectTx'
+
+import PayWarningModal from 'components/shared/PayWarningModal'
+
+import { readNetwork } from 'constants/networks'
+import { disablePayOverrides } from 'constants/v1/overrides'
+import { V1_PROJECT_IDS } from 'constants/v1/projectIds'
+import { V1_CURRENCY_ETH, V1_CURRENCY_USD } from 'constants/v1/currency'
```
Comment (aeolianeth) [inline src/components/v1/V1Project/Pay/V1PayButton.tsx:17]: **question**: couldn't this also use the new currencyContext now?

File: src/components/v1/V1Project/modals/V1ConfirmPayOwnerModal/V1ProjectRiskNotice.tsx [hunk 6]
```diff
@@ -13,7 +13,7 @@ import {
   FUNDING_CYCLE_WARNING_TEXT,
 } from 'constants/fundingWarningText'
 
-export default function ProjectRiskNotice() {
+export default function V1ProjectRiskNotice() {
```
Comment (aeolianeth) [inline src/components/v1/V1Project/modals/V1ConfirmPayOwnerModal/V1ProjectRiskNotice.tsx:16]: **praise** good catch

File: src/hooks/v1/transactor/PayV1ProjectTx.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
+import { NetworkContext } from 'contexts/networkContext'
+import { V1ProjectContext } from 'contexts/v1/projectContext'
+import { V1UserContext } from 'contexts/v1/userContext'
+
+import { useContext } from 'react'
+
+import { BigNumber } from 'ethers'
```
Comment (aeolianeth) [inline src/hooks/v1/transactor/PayV1ProjectTx.ts:7]: **suggestion**:

```suggestion
import { BigNumber } from '@ethersproject/bignumber'
```

File: src/hooks/v1/transactor/PayV1ProjectTx.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import { NetworkContext } from 'contexts/networkContext'
```
Comment (aeolianeth) [inline src/hooks/v1/transactor/PayV1ProjectTx.ts:1]: **question** is the only thing different in this file the name of the file?

**Review by aeolianeth — COMMENTED** [review]
Great great stuff @johnnyd-eth, left a few comments

File: src/hooks/v1/transactor/PayV1ProjectTx.ts (hunk 8 above)
Comment (johnnyd-eth) [reply src/hooks/v1/transactor/PayV1ProjectTx.ts:1]: Correct

File: src/components/v1/V1Project/Pay/V1PayButton.tsx (hunk 5 above)
Comment (johnnyd-eth) [reply src/components/v1/V1Project/Pay/V1PayButton.tsx:17]: I was thinking the currencyContext is only necessary for version-agnostic components, and that it might be simpler (at least for this PR) to continue using `V1_CURRENCY...` for version-specific components

File: src/components/shared/inputs/Pay/PayInputSubText.tsx (hunk 3 above)
Comment (johnnyd-eth) [reply src/components/shared/inputs/Pay/PayInputSubText.tsx:57]: Ok yeah good point, I've made a `useWeiConverter` hook which does the same thing as the previous (now deleted) `AmountToWei` util function, is this what you were thinking?

File: src/components/shared/inputs/Pay/PayInput.tsx (hunk 1 above)
Comment (johnnyd-eth) [reply src/components/shared/inputs/Pay/PayInput.tsx:72]: Good call

**Comment by johnnyd-eth** [comment]
Thanks for the comments @aeolianeth, replied / addressed 🙌

File: src/components/v1/V1Project/Pay/V1PayButton.tsx (hunk 5 above)
Comment (aeolianeth) [reply src/components/v1/V1Project/Pay/V1PayButton.tsx:17]: Yeah sweet that makes sense 👍

File: src/hooks/useWeiConverter.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
+import { useCurrencyConverter } from 'hooks/v1/CurrencyConverter'
```
Comment (aeolianeth) [inline src/hooks/useWeiConverter.ts:1]: **suggestion** to follow the existing naming convention, we should name this file `WeiConverter.ts`

**Review by aeolianeth — CHANGES_REQUESTED** [review]
@johnnyd-eth this is all looking great ser.

I've left one comment about typing. Not sure how big of a change this will require, but if it gets too big you could split this PR up into 2: 1 where we only add the currency context, and  where we add new pay button components.

Lemme know.

File: src/hooks/useWeiConverter.ts (hunk 9 above)
Comment (johnnyd-eth) [reply src/hooks/useWeiConverter.ts:1]: Fo sho

File: src/hooks/useWeiConverter.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+import { useContext } from 'react'
+import { CurrencyContext } from 'contexts/currencyContext'
+
+// Converts an amount (ETH or USD) to WEI
+export default function useWeiConverter({
+  currency,
+  amount,
+}: {
+  currency?: number
```
[the remark this answers is not in the document]
Comment (johnnyd-eth) [reply src/hooks/useWeiConverter.ts:12]: See message here for reasoning behind not addressing this: https://discord.com/channels/775859454780244028/951736957468282880/951798493683539968

tl;dr: opened up can of worms, am going to redo all of currency context anyway. Not worth blocking v2 pay button any longer IMO

File: src/utils/math.ts [hunk 11]
```diff
@@ -1,23 +1,25 @@
 import { BigNumber } from '@ethersproject/bignumber'
-import { V1FundingCycle } from 'models/v1/fundingCycle'
 
 import { fromWad, parsePerbicent } from './formatNumber'
-import { decodeFundingCycleMetadata } from './v1/fundingCycle'
 
 export const weightedRate = (
```
Comment (johnnyd-eth) [inline src/utils/math.ts:5]: Makes `weightedRate` version agnostic

File: src/contexts/currencyContext.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  V2_CURRENCY_METADATA,
+} from 'utils/v2/currency'
+
+import { CurrencyMetadata } from 'constants/currency'
+
+export type CurrencyMetadataType = Record<number, CurrencyMetadata>
+
+export type CurrencyContextType = {
+  currencyMetadata: CurrencyMetadataType
+  currencies: Record<string, number>
```
Comment (aeolianeth) [inline src/contexts/currencyContext.ts:14]: @johnnyd-eth can you pls create an issue to address this? Want to make sure we get to this eventually.

File: src/components/shared/PayWarningModal.tsx [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+      width={400}
+    >
+      <h2>
+        <Trans>Heads up</Trans>
+      </h2>
+      <p style={{ fontWeight: 500 }}>
+        <ExternalLink href="https://github.com/jbx-protocol/juice-contracts">
+          <Trans>Juicebox contracts</Trans>
+        </ExternalLink>{' '}
+        <Trans>
```
Comment (aeolianeth) [inline src/components/shared/PayWarningModal.tsx:30]: **suggestion**: let's wrap the whole text (including link) in a `Trans`, and remove the `Trans` on line 28. This way we're translating the whole sentence.

File: src/components/v1/V1Project/Pay/V1PayButton.tsx [hunk 14]
```diff
@@ -0,0 +1,137 @@
+import { t, Trans } from '@lingui/macro'
```
Comment (aeolianeth) [inline src/components/v1/V1Project/Pay/V1PayButton.tsx:1]: **suggestion**: given this is the only file in the `Pay` directory, can probably move outside of this directory, and delete it. WDYT?

File: src/components/v1/V1Project/index.tsx [hunk 15]
```diff
@@ -46,7 +55,13 @@ export default function V1Project({
         </Col>
 
         <Col xs={24} md={column ? 24 : 12} style={{ marginTop: gutter }}>
-          <Pay />
+          <PayInputGroup
```
Comment (aeolianeth) [inline src/components/v1/V1Project/index.tsx:58]: **praise** looks really great!

**Review by aeolianeth — CHANGES_REQUESTED** [review]
@johnnyd-eth 2 more tiny comments, then good to merge!

I tested pay functionality locally and it works great 👍

File: src/components/shared/PayWarningModal.tsx (hunk 13 above)
Comment (johnnyd-eth) [reply src/components/shared/PayWarningModal.tsx:30]: 100%

File: src/contexts/currencyContext.ts (hunk 12 above)
Comment (johnnyd-eth) [reply src/contexts/currencyContext.ts:14]: Done

File: src/components/v1/V1Project/Pay/V1PayButton.tsx (hunk 14 above)
Comment (johnnyd-eth) [reply src/components/v1/V1Project/Pay/V1PayButton.tsx:1]: Yep

**Comment by johnnyd-eth** [comment]
@aeolianeth thanks for that. Addressed your latest

**Review by aeolianeth — APPROVED** [review]
